File: domain/model/pixKey.go
package model

import (
	"errors"
	"github.com/asaskevich/govalidator"
	uuid "github.com/satori/go.uuid"
	"time"
)

type PixKey struct {
	Base `valid:"required"`
	Kind         string    `json:"kind" valid:"notnull"`
	Key          string    `json:"key" valid:"notnull"`
	AccountID    string    `json:"account_id" valid:"notnull"`
	Account      *Account  `valid:"-"`
	Status       string    `json:"status" valid:"notnull"`
}

type PixKeyRepository interface {
	RegisterKey(pixKey *PixKey) (*PixKey, error)
	FindKeyByKind(key string, kind string) (*PixKey, error)
	AddBank(bank *Bank) error
	AddAccount(account Account) error
	FindAccount(id string) (*Account, error)
}

func (pixKey *PixKey) isValid() error {
	_, err := govalidator.ValidateStruct(pixKey)

	if pixKey.kind != "email" && pixKey.kind != "cpf" {
		return errors.new("invalid type of key")
	}

	if pixKey.status != "active" && pixKey.status != "inactive" {
		return errors.new("invalid status")
	}
	
	if err != nil {
		return err
	}
	
	return nil
}

func NewPixKey(kind string, account *Account, key string) (*PixKey, error) {
	pixKey := PixKey{
		Kind: kind,
		Key: key,
		Account: account,
		Status: "active",
	}

	pixKey.ID = uuid.NewV4().String()
	pixKey.CreatedAt = time.Now()

	err := pixKey.isValid()

	if err != nil {
		return nil, err
	}

	return &pixKey, nil
}

File: codepix/domain/model/pix-key.js
import * as uuid from "uuid"

import { Base } from "./base.js";

export class PixKey extends Base {
  kind
  key
  accountId
  account
  status
  
  constructor({ kind, key, account, status }) {
    super()

    this.kind = kind
    this.key = key
    this.accountId = account.id
    this.account = account
    this.status = status
  }

  isValid() {
    if(this.kind !== "email" && this.kind !== "cpf") {
      return "'kind' must be of type 'email' or 'cpf'!"
    }

    if(this.status !== "active" && this.status !== "inactive") {
      return "'status' must be of type 'active' or 'inactive'!"
    }
    
    return null
  }

  static New(kind, account, key) {
    const pixKey = new PixKey({
      kind,
      account,
      key,
      status: "active"
    })

    pixKey.id = uuid.v4()
    pixKey.createdAt = new Date()

    const error = pixKey.isValid()

    if(error != null) {
      return [null, error]
    }

    return [pixKey, null]
  }
}